core/src/main/scala/org/http4s/multipart/MultipartParser.scala

```diff
@@ -4,153 +4,498 @@ package multipart
 
 /** A low-level multipart-parsing pipe.  Most end users will prefer EntityDecoder[Multipart]. */
 object MultipartParser {
 
-  private[this] val logger = org.log4s.getLogger
+  private val CRLFBytes = Array[Byte]('\r', '\n')
+  private val DoubleCRLFBytes = Array[Byte]('\r', '\n', '\r', '\n')
+  private val DashDashBytes = Array[Byte]('-', '-')
+  private val boundaryBytesR: Boundary => Array[Byte] = boundary => boundary.value.getBytes("UTF-8")
+  val startLineBytesR: Boundary => Array[Byte] = boundaryBytesR.andThen(DashDashBytes ++ _)
 
-  private val CRLFBytes = ByteVector('\r', '\n')
-  private val DashDashBytes = ByteVector('-', '-')
-  private val boundaryBytes: Boundary => ByteVector = boundary =>
-    ByteVector(boundary.value.getBytes)
-  private val startLineBytes: Boundary => ByteVector = boundaryBytes.andThen(DashDashBytes ++ _)
-  private val endLineBytes: Boundary => ByteVector = startLineBytes.andThen(_ ++ DashDashBytes)
-  private val expectedBytes: Boundary => ByteVector = startLineBytes.andThen(CRLFBytes ++ _)
+  private val expectedBytesR: Boundary => Array[Byte] =
+    boundaryBytesR.andThen(CRLFBytes ++ DashDashBytes ++ _)
 
-  final case class Out[+A](a: A, tail: Option[ByteVector] = None)
+  def parseStreamed[F[_]: Sync](
+      boundary: Boundary,
+      limit: Int = 1024): Pipe[F, Byte, Multipart[F]] = { st =>
+    ignorePreludeStage[F](boundary, st, limit)
+      .fold(Vector.empty[Part[F]])(_ :+ _)
+      .map(Multipart(_, boundary))
+  }
+
+  def parseToPartsStream[F[_]: Sync](
+      boundary: Boundary,
+      limit: Int = 1024): Pipe[F, Byte, Part[F]] = { st =>
+    ignorePreludeStage[F](boundary, st, limit)
+  }
 
-  def parse[F[_]: Sync](boundary: Boundary): Pipe[F, Byte, Either[Headers, ByteVector]] = s => {
-    val bufferedMultipartT = s.compile.toVector.map(ByteVector(_))
-    val parts = bufferedMultipartT.flatMap(parseToParts(_)(boundary))
-    val listT = parts.map(splitParts(_)(boundary)(List.empty[Either[Headers, ByteVector]]))
+  /** This function is the integral component in detecting whether `values`
+    * is either entirely contained in the chunk, partially contained
+    * in the chunk or not in the chunk at all
+    *
+    * @param values the value to look for in the chunk
+    * @param valueIx the current index of the value relative to the chunk.
+    *                This can mean the value is split across chunks.
+    * @param c the offending chunk.
+    */
+  private def stateIndex(values: Array[Byte], valueIx: Int, c: Chunk[Byte]): (Int, Int) = {
+    var i = 0
+    var sti = 0
+    val len2 = values.length - valueIx
+    while (sti < len2 && i < c.size) {
+      if (c(i) == values(sti + valueIx)) {
+        sti += 1
+      } else if (c(i) == values(0)) {
+        sti = 1
+      } else {
+        sti = 0
+      }
+      i += 1
+    }
+    (sti, i)
+  }
 
-    Stream
-      .eval(listT)
-      .flatMap(list => Stream.emits(list))
+  //Todo: This may fail for an edge case
```

Comment:
Will likely need more specifics about this.

Reply:
`stateIndex`  takes in as input a byte sequence (in an array), a chunk (which we can treat like an array, since we iterate through it) and a `state`, being an index to possibly the last place when this function was called. It returns the next index to check of the byte sequence to check, as well as the last place in the chunk where it ended.. A concrete example is, say we have two chunks like this, and the following byte sequence:
```scala
c1 = Chunk('a', 'b', 'c', 'c')
c2 = Chunk('d', 'e', 'f', 'z')

sequence = Array('c', 'd', 'e')
stateIndex(sequence, c1, 0) == (1, 4)
stateIndex(sequence, c2, 1) == (3, 2)
```

This is fine and dandy for sequences that perfectly align, but if we had:
```s
c1 = Chunk('a', 'b', 'c', 'c')
c2 = Chunk('a', 'c', 'd', 'e')

sequence = Array('c', 'd', 'e')
stateIndex(sequence, c1, 0) == (1, 4)
stateIndex(sequence, c2, 1) == (3, 4)
```

The problem here being that `splitAndIgnorePrev`, in the second case:
```scala
else if (i == c.size) {
    (sti + state, Stream.empty)
````

so for our hypothetical chunk, it considers the _entire_ chunk as part of the sequence to ignore, but only the last 3 bytes were part of it

So I'm considering that this function, in a very specific scenario can fail, but I haven't been able to make the perfect storm happen.

I need to replace this function specifically, but otherwise everything else about this is ok.